README
NAME

    Plugin::Simple - Load plugins from files or modules.

SYNOPSIS

        use Plugin::Simple;
    
        # load a plugin module from a file
    
        @plugins = plugins('/path/to/MyModule.pm');
    
        # load all modules under '__PACKAGE__::Plugin' namespace
    
        my @plugins = plugins(); # call in scalar context to retrieve the first one
    
        # load all plugins under a specific namespace (note the trailing ::)
    
        @plugins = plugins('Any::Namespace::');
    
        # load/return only the plugins that can perform specific functions
    
        @plugins = plugins(can => ['foo', 'bar]); # foo and bar
    
        # instead of importing 'plugins()', change the name:
    
        use Plugin::Simple sub_name => 'foo';
        @plugins = foo(...);
    
        # do something with the plugins
    
        for my $plugin (@plugins){
            $plugin->plugin_func(@args);
        }
    
        # works in OO modules too simply by using it
    
        my @plugins = $self->plugins();

DESCRIPTION

    There are many plugin modules available on the CPAN, but I wrote this
    one just for fun. It's very simple, extremely lightweight, and is
    extremely minimalistic in what it does.

    It searches for modules in installed packages or non-installed files,
    and loads them (without string eval). You can optionally have us return
    only the plugins that can() perform a specific task.

FUNCTIONS/METHODS

    None. We simply install a plugin() function within the namespace of the
    package that used us. To specify a different sub name, use this module
    as such: use Plugin::Simple sub_name => 'name_of_sub';.

AUTHOR

    Steve Bertrand, <steveb at cpan.org>

 CONTRIBUTING

    Any and all feedback and help is appreciated. A Pull Request is the
    preferred method of receiving changes
    (https://github.com/stevieb9/p5-plugin-simple), but regular patches
    through the bug tracker, or even just email discussions are welcomed.

BUGS

    https://github.com/stevieb9/p5-plugin-simple/issues

SUPPORT

    You can find documentation for this script and module with the perldoc
    command.

        perldoc Plugin::Simple;

SEE ALSO

    There are far too many plugin import modules on the CPAN to mention
    here.

LICENSE AND COPYRIGHT

    Copyright 2016 Steve Bertrand.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Plugin-Simple

1.02 UNREL
        -

1.01    2018-02-11
        - remove perl 5.8 from Travis-CI due to coverage failing to install

1.00    2018-02-11
        - fix missing closing single quote in SYNOPSIS (closes #10)
        - remove CONTRIBUTING section in POD (closes #9)
        - fix incorrect sub name in POD and warnings (closes #7)
        - added example use case in EXAMPLE (closes #8)

0.07    2016-03-28
        - fixed a missed croak() if a plugin file can't be found, breaking
          tests on certain MSWin platforms

0.06    2016-03-25
        - we now croak if a plugin package can't be found and no default is
          set (closes #4)
        - package in cwd is now loaded properly (closes #5)


0.05    2016-03-22
        - removed _cache()
        - removed _config()
        - updated POD to reflect 'default' option
        - fixed bug where default plugin wasn't being loaded before can() was 
          being tested


0.04    2016-03-21
        - in scalar context, plugins() will return the first plugin found
        - we now use Module::Load to do the dirty work, as ExtUtils::Installed
          relies on .packlist files, which not all modules provide
        - changed certain tests to use local modules instead of skipping
        - added github links into Makefile.PL (thanks Mohammad Anwar (MANWAR))

0.03    2016-03-21
        - fixed github links
        - cleaned up POD errors

0.02    2016-03-21
        - added 'can => [sub1, sub2]' param to plugin()
        - ability to load plugins from a file
        - test coverage 100%

0.01    2016-03-21
        First version, released on an unsuspecting world.



t/base/Testing.pm
package Testing;
use strict;
use warnings;

sub hello {
    return "hello, world!";
}

1;

t/04-plugins_sub.pm
#!/usr/bin/perl
use strict;
use warnings;

use Plugin::Simple cache => 1;
use Test::More;

my $file = 't/base/Testing.pm';

{ # file
    my @plugins = plugins($file);
    is (@plugins, 1, "with file, got the correct number of plugins");
    is ($plugins[0], 'Testing', "...and the plugin name is ok");
    is ($plugins[0]->hello(), 'hello, world!', "...and the sub output ok");
}
SKIP: { # module test
    my @ret = plugins('DateTime');

    skip 'test mod not installed', 2, unless defined $INC{'DateTime.pm'};
    is (@ret, 5, "with an example item, things appear ok");
    is ($ret[0], 'DateTime', 'returned plugin is correct');
    can_ok('DateTime', 'new');
};
{ # can param bad
    my @plugins = plugins($file, can => 'blah');
    is (@plugins, 0, "if plugin can()t, it isn't returned");
}
{ # can param good
    my @plugins = plugins($file, can => 'hello');
    is (@plugins, 1, "if plugin can(), it's returned");
}
{ # can w/o item
    my @plugins = plugins(can => 'hello');
    is (@plugins, 0, "'can' can be sent in without item");
}
{ # can with wrong prototype
    my $ok = eval { plugins(can => 'hello', 'this'); 1; };
    ok (! $ok, "with wrong order of params, plugins() croaks");
}
done_testing;


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Plugin::Simple',
    AUTHOR           => q{Steve Bertrand <steveb@cpan.org>},
    VERSION_FROM     => 'lib/Plugin/Simple.pm',
    ABSTRACT_FROM    => 'lib/Plugin/Simple.pm',
    LICENSE          => 'perl_5',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
    },
    PREREQ_PM => {
        'Module::Load' => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Plugin-Simple-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/stevieb9/p5-plugin-simple.git',
                web  => 'https://github.com/stevieb9/p5-plugin-simple',
            },
        }})
     : ()
    ),
);


lib/Module/Plugin/Import.pm
package Module::Plugin::Import;

use 5.006;
use strict;
use warnings;

use Data::Dumper;
use ExtUtils::Installed;
use Module::Load;

sub import {
    my $class = shift;
    my %opts = @_;

sub plugins {
    my $from = shift;
    
    my $inst = ExtUtils::Installed->new;
    my @modules = $inst->modules;

    my $pkg = _caller();
    $plugin_loc = _from($from) ? _from($from) : "$pkg::Plugin::";

    my @plugins = grep /^$plugin_loc/, @modules;

    for (@plugins){
        load $_;
    }
}
sub _caller {
    my ($pkg, $file) = caller;
    return $pkg;
}
sub _from {
    my $loc = shift;
    if ($loc){
        $loc .= ($loc =~ /::$/) ? '' : '::';
    }
    return $loc;
}
sub _export_symbols {

=head1 NAME

Module::Plugin::Import - The great new Module::Plugin::Import!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Module::Plugin::Import;

    my $foo = Module::Plugin::Import->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub function1 {
}

=head2 function2

=cut

sub function2 {
}

=head1 AUTHOR

Steve Bertrand, C<< <steveb at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-module-plugin-import at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Module-Plugin-Import>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Module::Plugin::Import


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Plugin-Import>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Module-Plugin-Import>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Module-Plugin-Import>

=item * Search CPAN

L<http://search.cpan.org/dist/Module-Plugin-Import/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2016 Steve Bertrand.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut

1; # End of Module::Plugin::Import


lib/Plugin/Simple.pm
package Plugin::Simple;
use 5.006;
use strict;
use warnings;

use Carp qw(croak);
use Cwd qw (abs_path);
use Module::List qw(list_modules);
use Module::Load;

our $VERSION = '1.02';

my $self;

sub import {
    my ($class, %opts) = @_;

    $self = __PACKAGE__->_new(%opts);

    my $sub_name = $opts{sub_name} ? $opts{sub_name} : 'plugins';

    {
        no warnings 'redefine';
        no strict 'refs';

        my $pkg = (caller)[0];
        *{"$pkg\::$sub_name"} = \&_plugins;
    }
}
sub _new {
    my ($class, %args) = @_;
    my $self = bless \%args, $class;

    return $self;
}
sub _search {
    my ($self, $pkg, $item) = @_;

    my @plugins;

    if ($item){
        if ($item !~ /::$/){
            push @plugins, $item;
        }
        else {
            my $candidates;
            eval { $candidates = list_modules(
                    $item,
                    {list_modules => 1, recurse => 1}
                );
            };
            push @plugins, keys %$candidates;
        }
    }
    else {
        my $path = $pkg;
        $path .= '::Plugin::';
        my $candidates = {};
        eval { $candidates = list_modules(
                $path,
                {
                    list_modules => 1,
                    recurse => 1
                }
            );
        };
        push @plugins, keys %$candidates;
    }

    my @loaded;

    for (@plugins){
        my $ok = $self->_load($_);
        push @loaded, $ok;
    }

    return @loaded;
}
sub _load {
    my ($self, $plugin) = @_;

    if ($plugin =~ /(.*)\W(\w+)\.pm/){
        unshift @INC, $1;
        $plugin = $2;
    }
    elsif ($plugin =~ /^(\w+)\.pm$/){
        unshift @INC, '.';
        $plugin = $1;
    }

    my $loaded = eval { load $plugin; 1; };

    if ($loaded) {
        return $plugin;
    }
}
sub _plugins {
    shift if ref $_[0]; # dump the calling object if present

    my ($item, $can);

    if ($_[0] && $_[0] eq 'can'){
        shift;
        $can = shift;
    }
    else {
        $item = shift;
        shift;
        $can = shift;
    }

    if (@_){
        croak "usage: plugins(['Load::From'], [can => 'sub']), " .
              "in that order\n";
    }

    my $pkg = (caller)[0];
    my @plugins;

    if ($item){
        if ($item =~ /(?:\.pm|\.pl)/){
            my $abs_path;
            my $ok_file = eval { $abs_path = abs_path($item); 1 };

            if (! $ok_file){
                croak
                "\npackage $item can't be found, and no default plugin set\n";
            }

            if (-e $abs_path){
                @plugins = $self->_load($abs_path);
            }
        }
        else{
            @plugins = $self->_search($pkg, $item);
        }
    }
    if (! @plugins){
        @plugins = _search($pkg);
    }
    if (! $plugins[0] && $self->{default}){
        push @plugins, $self->_load($self->{default});
    }
    if (! $plugins[0]){
        if ($item){
            croak
            "\npackage $item can't be found, and no default plugin set\n";
        }
        else {
            croak "\npackage can't be found, and no default plugin set\n";
        }
    }
    my @wanted_plugins;

    if ($can) {
        for my $mod (@plugins){
            my $can_count = 0;
            for my $sub (@$can){
                if ($mod->can($sub)){
                    $can_count++;
                }
            }
            push @wanted_plugins, $mod if $can_count == @$can;
        }
        return wantarray ? @wanted_plugins : $wanted_plugins[0];
    }

    return wantarray ? @plugins : $plugins[0];
}

1;

=head1 NAME

Plugin::Simple - Load plugins from files or modules.

=for html
<a href="http://travis-ci.org/stevieb9/p5-plugin-simple"><img src="https://secure.travis-ci.org/stevieb9/p5-plugin-simple.png"/>
<a href='https://coveralls.io/github/stevieb9/p5-plugin-simple?branch=master'><img src='https://coveralls.io/repos/stevieb9/p5-plugin-simple/badge.svg?branch=master&service=github' alt='Coverage Status' /></a>

=head1 SYNOPSIS

    use Plugin::Simple;

    # load a plugin module from a file

    @plugins = plugins('/path/to/MyModule.pm');

    # load all modules under '__PACKAGE__::Plugin' namespace

    my @plugins = plugins(); # call in scalar context to retrieve the first one

    # load all plugins under a specific namespace (note the trailing ::)

    @plugins = plugins('Any::Namespace::');

    # load/return only the plugins that can perform specific functions

    @plugins = plugins(can => ['foo', 'bar']); # foo and bar

    # instead of importing 'plugins()', change the name:

    use Plugin::Simple sub_name => 'foo';
    @plugins = foo(...);

    # set a default fallback plugin if searching turns up nothing

    use Plugin::Simple default => 'My::Module::Plugin::DefaultPlugin'

    # do something with the plugins

    for my $plugin (@plugins){
        $plugin->plugin_func(@args);
    }

    # works in OO modules too simply by using it

    my @plugins = $self->plugins();

=head1 DESCRIPTION

There are many plugin modules available on the CPAN, but I wrote this one just
for fun. It's very simple, extremely lightweight, and is extremely minimalistic
in what it does.

It searches for modules in installed packages or non-installed files, and loads
them (without string C<eval>). You can optionally have us return only the
plugins that C<can()> perform a specific task.

=head1 LOAD OPTIONS

By default, we force C<plugins()> into your namespace. To change this name:

    use Plugin::Simple sub_name => 'other_name';

If searching fails, you can ensure a default known plugin gets loaded:

    use Plugin::Simple default => 'My::Plugin';

To use both options, simply separate them with a comma.

=head1 FUNCTIONS/METHODS

None. We simply install a C<plugins()> function within the namespace of the
package that C<use>d us.

=head1 EXAMPLE

This example simply uses a single plugin module with a C<plugin_function()>
function. In the script, we load this file, and check to ensure the plugin does
in fact have that sub available.

We then call the plugins in a loop (even though in this case there's only one),
and send in an argument for the plugin to do work on.

=head2 Script

    use warnings;
    use strict;

    use lib '.';

    use Plugin::Simple;

    my @plugins = plugins(
        'examples/TestPlugin.pm',
        can => ['plugin_function']
    );

    my $plugin_arg = 'Hello!';

    for my $plugin (@plugins){
        $plugin->plugin_function($plugin_arg);
    }

=head2 Plugin Module

    package TestPlugin;

    sub plugin_function {
        shift; # throw away class/obj
        my ($str) = @_;
        print "in " . __PACKAGE__ . ", arg is: $str\n";
    }

    1;

=head2 Output

    in TestPlugin, arg is: Hello!

=head1 AUTHOR

Steve Bertrand, C<< <steveb at cpan.org> >>

=head1 BUGS

L<https://github.com/stevieb9/p5-plugin-simple/issues>

=head1 SEE ALSO

There are far too many plugin import modules on the CPAN to mention here.

=head1 LICENSE AND COPYRIGHT

Copyright 2016,2017,2018 Steve Bertrand.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

